### PR TITLE
feat: Validate headers against validation expression property

### PR DESCRIPTION
**Pull request by lucashuy**
#### Which issue(s) does this change fix?

N/A

#### Why is this change necessary?
Adds a missing check to validate incoming headers against the validation expression property defined in the template file.

#### How does it address the issue?
Modifies the existing header validator by overriding the parent validation method to also check the expression.

#### What side effects does this change have?
None

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by mildaniel — COMMENTED** [review]
Thanks Lucas, LGTM. I'll defer to Mohamed's comment about testing and approve once that's complete.

**Review by mildaniel — APPROVED** [review]

**Review by moelasmar — APPROVED** [review]

**Pull request merged**
